# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/workers/RuntimeService.cpp
# Commit: d995d3771ad5
# Full Hash: d995d3771ad549f75d9f53da7bd62a203ae95ac8
# Author: Kris Maglione <maglione.k@gmail.com>
# Date: 2020-04-30 08:26:21
# Regressor Bug: 1594572
# File Overlap Count: 1
# Description:
#   Bug 1594572: Schedule worker deletion when primary runnable fails. r=asuth
#   
#   If we try to start a worker too close to shutdown, the main runnable for its
#   thread runs after shutdown has begun, bails out early trying to create a
#   BackgroundChild for its thread. Unlike bail-outs in WorkerPrivate::DoRunLoop,
# ==============================================================================

diff -r 244a33bcd5e7 -r d995d3771ad5 dom/workers/RuntimeService.cpp
--- a/dom/workers/RuntimeService.cpp	Wed Apr 29 16:40:55 2020 +0000
+++ b/dom/workers/RuntimeService.cpp	Thu Apr 30 03:52:56 2020 +0000
@@ -2235,6 +2235,13 @@
 
   SetThreadHelper threadHelper(mWorkerPrivate, mThread);
 
+  auto failureCleanup = MakeScopeExit([&]() {
+    // The creation of threadHelper above is the point at which a worker is
+    // considered to have run, because the `mPreStartRunnables` are all
+    // re-dispatched after `mThread` is set.
+    mWorkerPrivate->ScheduleDeletion(WorkerPrivate::WorkerRan);
+  });
+
   mWorkerPrivate->AssertIsOnWorkerThread();
 
   // These need to be initialized on the worker thread before being used on
@@ -2265,6 +2272,8 @@
       return NS_ERROR_FAILURE;
     }
 
+    failureCleanup.release();
+
     {
       PROFILER_SET_JS_CONTEXT(cx);
 